File: Lab 13/13-1.py
import heapq

class Solution(object):
    def minCostConnectPoints(self, points):
    
        n = len(points)
        visited = [False] * n
        min_heap = [(0, 0)] 
        total_cost = 0
        edges_used = 0

        def manhattan_dist(p1, p2):
            return abs(p1[0] - p2[0]) + abs(p1[1] - p2[1])

        while edges_used < n:
            cost, curr_point = heapq.heappop(min_heap)
            if visited[curr_point]:
                continue

            visited[curr_point] = True
            total_cost += cost
            edges_used += 1

            for next_point in range(n):
                if not visited[next_point]:
                    dist = manhattan_dist(points[curr_point], points[next_point])
                    heapq.heappush(min_heap, (dist, next_point))

        return total_cost

File: Lab 13/13-1.java
import java.util.PriorityQueue;

class Solution {
    public int minCostConnectPoints(int[][] points) {
        int n = points.length;

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);

        boolean[] visited = new boolean[n];

        int totalCost = 0;
        int edgesUsed = 0;

        pq.offer(new int[] { 0, 0 });

        while (edgesUsed < n) {
            int[] current = pq.poll();
            int cost = current[0];
            int point = current[1];

            if (visited[point]) {
                continue;
            }

            visited[point] = true;

            totalCost += cost;
            edgesUsed++;

            for (int next = 0; next < n; next++) {
                if (!visited[next]) {
                    int manhattanDistance = Math.abs(points[point][0] - points[next][0]) +
                            Math.abs(points[point][1] - points[next][1]);
                    pq.offer(new int[] { manhattanDistance, next });
                }
            }
        }

        return totalCost;
    }
}
